### ISSUE TITLE
More flexible rendering of quest markers for linear features

### ISSUE THREAD

**Issue by althio** [westnordost/StreetComplete#733]
This is about markers for quests related to linear features, typically roads surface, street names, ...

Roads segments can be quite long and _it looks like_ the marker is located at a fixed position on the middle-point/centroid of the linear features.
_I think_ that at the typical zoom used in the app, you can easily walk on a street or cross a street but miss a marker. This is because the marker quest is out of the screen.

Suggestions:
- add several markers, regularly spaced on the line (space may be zoom-dependant)
- see if the marker position can be floating, anywhere on the line, to dynamically choose the position on the line, within the screen, without colliding too much with others markers/labels.

**Comment by westnordost** [comment]
Yes, it is placed on the centroid.

Your seconds suggestion is technically not possible, regarding your first suggestion: 
It would be possible but I don't really like it. People would get confused because it looks as if there were several quests when there is only one.

Note that most linear quests require you to walk down the whole of the road, so theoretically, one can not miss it.

Am 21. Dezember 2017 17:11:03 GMT+07:00 schrieb "Benoît" <notifications@github.com>:
>This is about markers for quests related to linear features, typically
>roads surface, street names, ...
>
>Roads segments can be quite long and _it looks like_ the marker is
>located at a fixed position on the middle-point/centroid of the linear
>features.
>_I think_ that at the typical zoom used in the app, you can easily walk
>on a street or cross a street but miss a marker. This is because the
>marker quest is out of the screen.
>
>Suggestions:
>- add several markers, regularly spaced on the line (space may be
>zoom-dependant)
>- see if the marker position can be floating, anywhere on the line, to
>dynamically choose the position on the line, within the screen, without
>colliding too much with others markers/labels.
>
>-- 
>You are receiving this because you are subscribed to this thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/westnordost/StreetComplete/issues/733

-- 
Diese Nachricht wurde von meinem Android-Mobiltelefon mit K-9 Mail gesendet.

**Comment by esilja** [comment]
Could you expand why the second suggestion is technically impossible?

**Comment by althio** [comment]
>> add several markers, regularly spaced on the line (space may be zoom-dependant)
> regarding your first suggestion: It would be possible but I don't really like it. People would get confused because it looks as if there were several quests when there is only one.

Is it possible to choose the spacing so that marker could appear only once on screen? (but at least marker position is not tied to centroid only)

> Note that most linear quests require you to walk down the whole of the road, so theoretically, one can not miss it.

That is not my experience.
A street name is more easily found at the beginning or the end, not at the centroid, and I don't think you are required to walk down the whole street.
Also if you are walking with GPS on, at high-zoom, you can cross streets/roads and never know they have quests to begin with. Kind of defeat the purpose of walking them down.

**Comment by westnordost** [comment]
@eslija: For the map rendering, I use a third party library which does not allow anything like that. To be honest, the current functionality is already a bit of a hack.

Am 21. Dezember 2017 20:06:57 GMT+07:00 schrieb esilja <notifications@github.com>:
>Could you expand why the second suggestion is technically impossible?
>
>-- 
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/westnordost/StreetComplete/issues/733#issuecomment-353346181

-- 
Diese Nachricht wurde von meinem Android-Mobiltelefon mit K-9 Mail gesendet.

**Comment by westnordost** [comment]
But you are still required to check if the whole stretch indicated has the same name. Street names may change at intersections.

> Is it possible to choose the spacing so that marker could appear only once on screen? (but at least marker position is not tied to centroid only)

No.

There is another ticket somewhere with a feature request somewhere in which the closest quest should be shown at the border of the screen and I plan to implement it as soon as it becomes technically possible.

Am 21. Dezember 2017 20:09:16 GMT+07:00 schrieb "Benoît" <notifications@github.com>:
>>> add several markers, regularly spaced on the line (space may be
>zoom-dependant)
>> regarding your first suggestion: It would be possible but I don't
>really like it. People would get confused because it looks as if there
>were several quests when there is only one.
>
>Is it possible to choose the spacing so that marker could appear only
>once on screen? (but at least marker position is not tied to centroid
>only)
>
>> Note that most linear quests require you to walk down the whole of
>the road, so theoretically, one can not miss it.
>
>That is not my experience.
>A street name is more easily found at the beginning or the end, not at
>the centroid, and I don't think you are required to walk down the whole
>street.
>Also if you are walking with GPS on, at high-zoom, you can cross
>streets/roads and never know they have quests to begin with. Kind of
>defeat the purpose of walking them down.
>
>-- 
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/westnordost/StreetComplete/issues/733#issuecomment-353346689

-- 
Diese Nachricht wurde von meinem Android-Mobiltelefon mit K-9 Mail gesendet.

**Comment by althio** [comment]
Thank you for the technical answers. (I hope you still enjoy your travel!)
I will try some other replies and suggestions, let's see if they are useful.

> But you are still required to check if the whole stretch indicated has the same name. Street names may change at intersections.

"required to check if the whole stretch": I don't think that one ponctual marker (the centroid or any other point) is suitable to express this. Anyway I think that is clear enough when you click on the marker: app zooms out, show&highlight the whole feature. This is not an argument for centroid position versus other postion in my opinion.

On the other hand, I think the discoverability of the marker and the quest is not very good. If you don't see it, you cannot click on it.
I would rather have markers at near-beginning and near-end of the line than centroid for quest like _street names_ and _speed limit_. That is the natural place for street signs and speed signs.
Right now:
If you are at the centroid, you have the question but not the answer.
If you are at the extremity, you have the answer but not the question.
Distance between centroid and extremity can be large enough (pan several times screen size or zoom out a lot)

> People would get confused because it looks as if there were several quests when there is only one.

Really?
Even if the related markers and line are all shown & highlighted together on click?
And of course all removed together if quest is done or hidden?

> There is another ticket somewhere with a feature request somewhere in which the closest quest should be shown at the border of the screen and I plan to implement it as soon as it becomes technically possible.

This is https://github.com/westnordost/StreetComplete/issues/642#issuecomment-336106926 or https://github.com/westnordost/StreetComplete/issues/142#issuecomment-297359103 but this is only intented if no quest at all are shown.
It might mitigate the issue but will not resolve it in dense urban area where a lot of other quests (opening hours, ...) might appear between the extremity and the centroid of the street and block the discoverability of the quest related to the whole street.

**Comment by westnordost** [comment]
Jup, thanks. Currently I am in Northern Thailand :-) 

Hmm yes, perhaps one would need to simply try it out and see how the UX is then on a survey.

Am 22. Dezember 2017 17:17:28 GMT+07:00 schrieb "Benoît" <notifications@github.com>:
>Thank you for the technical answers. (I hope you still enjoy your
>travel!)
>I will try some other replies and suggestions, let's see if they are
>useful.
>
>> But you are still required to check if the whole stretch indicated
>has the same name. Street names may change at intersections.
>
>"required to check if the whole stretch": I don't think that one
>ponctual marker (the centroid or any other point) is suitable to
>express this. Anyway I think that is clear enough when you click on the
>marker: app zooms out, show&highlight the whole feature. This is not an
>argument for centroid position versus other postion in my opinion.
>
>On the other hand, I think the discoverability of the marker and the
>quest is not very good. If you don't see it, you cannot click on it.
>I would rather have markers at near-beginning and near-end of the line
>than centroid for quest like _street names_ and _speed limit_. That is
>the natural place for street signs and speed signs.
>Right now:
>If you are at the centroid, you have the question but not the answer.
>If you are at the extremity, you have the answer but not the question.
>Distance between centroid and extremity can be large enough (pan
>several times screen size or zoom out a lot)
>
>> People would get confused because it looks as if there were several
>quests when there is only one.
>
>Really?
>Even if the related markers and line are all shown & highlighted
>together on click?
>And of course all removed together if quest is done or hidden?
>
>> There is another ticket somewhere with a feature request somewhere in
>which the closest quest should be shown at the border of the screen and
>I plan to implement it as soon as it becomes technically possible.
>
>This is
>https://github.com/westnordost/StreetComplete/issues/642#issuecomment-336106926
>or
>https://github.com/westnordost/StreetComplete/issues/142#issuecomment-297359103
>but this is only intented if no quest at all are shown.
>It might mitigate the issue but will not resolve it in dense urban area
>where a lot of other quests (opening hours, ...) might appear between
>the extremity and the centroid of the street and block the
>discoverability of the quest related to the whole street.
>
>
>
>-- 
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/westnordost/StreetComplete/issues/733#issuecomment-353567979

-- 
Diese Nachricht wurde von meinem Android-Mobiltelefon mit K-9 Mail gesendet.

**Comment by westnordost** [comment]
I recently tried out StreetComplete in Prachuap Khiri Khan and I understand now that this feature can make sense. (But not for all linear quests.)

**Comment by westnordost** [comment]
Alright, implementation is almost done. I implemented it this way that the quest markers appear 15m away from the end of the element so that they are not all directly at the intersection. (So if the way is shorter than 4*15m, the quest marker is shown in the center as before)

The question is for which quests the markers should show at the ends. Definitely for oneway, maxspeed, road name and road construction because they all require to look for a sign at the entrance/exit of the road.

Currently I am thinking that the other way-based quests where the task is not to read a sign at the start/end of the road but to check something for the whole way (cycleway, cycleway segregation, surface, track type and way lit) should continue to show the quest marker in the center.

**Comment by rugk** [comment]
> The question is for which quests the markers should show at the ends

At _both_ ends?

If yes: I suppose it looks quite weird if you have a way, where you can see both ends and it looks as if you have a quest twice, or, if you solve it, you may wonder why another one disappears.
If not: I'd say if both ends of the street are visible, show the quest in the middle of the street, as before.

(or some kind of "floating" as it has been proposed before)

**Comment by westnordost** [comment]
Yes, at both ends. I think it will not be confusing because when you zoom into the quest, you clearly see that there is a quest marker for the same quest at both ends.

>  If not: I'd say if both ends of the street are visible, show the quest in the middle of the street, as before.

This is not possible.

**Comment by georg-d** [comment]
@westnordost Was the quest marker also defined to appear at beginning & end for road surface quest? Or was this lost at some point in version history? Background for the question: I repeatedly experience that I drive along a road, I do not see any quest marker, then a quest marker for e.g. road surface or smoothness becomes visible but I cannot answer properly because I did not pay attention to the asked property during the 500m before – if quest marker appeared at the beginning of the road segment, I had known what I shall pay attention to.

Example: My current position is quite away from the beginning of the road segment and only then, at the very top, the quest marker slowly becomes visible – in real life, I did not yet see it because I had zoomed in a little more <img width="400" height="800" alt="Image" src="https://github.com/user-attachments/assets/9400bf76-2bbe-4324-bff0-3a8550b86c50" />
